File: challenges/Algorithmic/Towers of Hanoi/ToH.py
"""ToH.py - Towers of Hanoi solver utilities (recursive & iterative).

Provides:
  * Recursive move generator (`towers_of_hanoi`)
  * Iterative solver (`towers_of_hanoi_iterative`) using an explicit stack
  * State generator (`hanoi_state_generator`) for visualization
  * CLI with JSON output, count-only mode, iterative toggle, and step limiting

All functions are pure (no printing) enabling reuse in teaching, testing, or
visualization contexts. Complexity is O(2^n) moves; memory for iterative stack
is O(n).
"""

from __future__ import annotations

import argparse
import json
import sys
from dataclasses import dataclass
from typing import Dict, Generator, Iterable, Iterator, List, Optional


Move = str


def towers_of_hanoi(
    disks: int, source: str, target: str, auxiliary: str
) -> Iterator[Move]:
    """Generates the sequence of moves to solve the Towers of Hanoi puzzle.

    This function uses a recursive approach to generate the moves.

    Args:
        disks: The number of disks to move.
        source: The name of the source peg.
        target: The name of the target peg.
        auxiliary: The name of the auxiliary peg.

    Yields:
        A string representing the next move in the sequence.
    """
    if disks > 0:
        yield from towers_of_hanoi(disks - 1, source, auxiliary, target)
        yield f"Move disk {disks} from {source} to {target}"
        yield from towers_of_hanoi(disks - 1, auxiliary, target, source)


def towers_of_hanoi_iterative(
    disks: int, source: str, target: str, auxiliary: str
) -> Iterator[Move]:
    """Generates the sequence of moves to solve the Towers of Hanoi puzzle.

    This function uses an iterative approach with an explicit stack to avoid
    recursion depth issues with a large number of disks.

    Args:
        disks: The number of disks to move.
        source: The name of the source peg.
        target: The name of the target peg.
        auxiliary: The name of the auxiliary peg.

    Yields:
        A string representing the next move in the sequence.
    """
    stack: List[tuple[int, str, str, str, int]] = [
        (disks, source, target, auxiliary, 0)
    ]
    while stack:
        n, src, tgt, aux, stage = stack.pop()
        if n == 0:
            continue
        if stage == 0:
            # Defer processing current move after left subtree
            stack.append((n, src, tgt, aux, 1))
            stack.append((n - 1, src, aux, tgt, 0))
        else:
            # Emit move then process right subtree
            yield f"Move disk {n} from {src} to {tgt}"
            stack.append((n - 1, aux, tgt, src, 0))


def hanoi_state_generator(
    num_disks: int, source: str, target: str, auxiliary: str
) -> Iterator[Dict[str, List[int]]]:
    """Generates the state of the pegs after each move.

    This is useful for creating visualizations of the puzzle.

    Args:
        num_disks: The number of disks.
        source: The name of the source peg.
        target: The name of the target peg.
        auxiliary: The name of the auxiliary peg.

    Yields:
        A dictionary representing the state of the pegs, where the keys are
        the peg names and the values are lists of the disks on each peg.
    """
    pegs: Dict[str, List[int]] = {
        source: list(range(num_disks, 0, -1)),
        auxiliary: [],
        target: [],
    }
    yield {k: v.copy() for k, v in pegs.items()}

    def solve(n: int, src: str, tgt: str, aux: str) -> Iterator[None]:
        if n > 0:
            yield from solve(n - 1, src, aux, tgt)
            disk = pegs[src].pop()
            pegs[tgt].append(disk)
            yield None
            yield from solve(n - 1, aux, tgt, src)

    for _ in solve(num_disks, source, target, auxiliary):
        yield {k: v.copy() for k, v in pegs.items()}


@dataclass(slots=True)
class CLIConfig:
    disks: int
    iterative: bool = False
    json_output: bool = False
    count_only: bool = False
    max_steps: Optional[int] = None

    def validate(self) -> None:
        if self.disks < 1:
            raise ValueError("disks must be >= 1")
        if self.max_steps is not None and self.max_steps < 1:
            raise ValueError("max-steps must be positive")


def solve_moves(cfg: CLIConfig) -> List[Move]:
    gen = (towers_of_hanoi_iterative if cfg.iterative else towers_of_hanoi)(
        cfg.disks, "A", "C", "B"
    )
    moves: List[Move] = []
    for i, mv in enumerate(gen, 1):
        moves.append(mv)
        if cfg.max_steps is not None and i >= cfg.max_steps:
            break
    return moves


def build_parser() -> argparse.ArgumentParser:
    p = argparse.ArgumentParser(description="Towers of Hanoi solver")
    p.add_argument("--disks", "-n", type=int, default=4, help="Number of disks (>=1)")
    p.add_argument(
        "--iterative", action="store_true", help="Use iterative solver implementation"
    )
    p.add_argument("--json", action="store_true", help="Emit JSON output")
    p.add_argument("--count-only", action="store_true", help="Only output move count")
    p.add_argument(
        "--max-steps", type=int, help="Limit number of moves emitted (for preview)"
    )
    return p


def main(argv: Optional[Iterable[str]] = None) -> int:
    parser = build_parser()
    args = parser.parse_args(list(argv) if argv is not None else None)
    cfg = CLIConfig(
        disks=args.disks,
        iterative=args.iterative,
        json_output=args.json,
        count_only=args.count_only,
        max_steps=args.max_steps,
    )
    try:
        cfg.validate()
    except ValueError as e:
        print(f"Error: {e}", file=sys.stderr)
        return 1

    total_moves = (2**cfg.disks) - 1
    moves = [] if cfg.count_only else solve_moves(cfg)

    if cfg.json_output:
        print(
            json.dumps(
                {
                    "disks": cfg.disks,
                    "iterative": cfg.iterative,
                    "total_moves": total_moves,
                    "emitted_moves": len(moves),
                    "moves": moves[
                        :1000
                    ],  # safety cap for extremely large n with max-steps
                },
                indent=2,
            )
        )
    else:
        if cfg.count_only:
            print(f"Total moves: {total_moves}")
        else:
            for i, mv in enumerate(moves, 1):
                print(f"{i}. {mv}")
            print(f"\nTotal moves required: {total_moves}")
    return 0


if __name__ == "__main__":  # pragma: no cover
    raise SystemExit(main())

File: challenges/Algorithmic/Towers of Hanoi/ToH.hs
{-# LANGUAGE RecordWildCards #-}

-- |
-- Module      : ToH
-- Description : Towers of Hanoi solvers (recursive and iterative) with CLI/JSON helpers.
--
-- The implementation mirrors the Python utilities in ToH.py, offering both
-- recursive and stack-based generators alongside a peg-state stream for
-- visualization integrations.  The executable exposes flags for selecting the
-- solver, limiting output, exporting JSON, and tracing intermediate peg
-- configurations.  All pure helpers remain easily reusable from GHCi or other
-- Haskell modules.
module Main
  ( Move(..)
  , PegState
  , towersOfHanoi
  , towersOfHanoiIterative
  , hanoiStateGenerator
  , CLIConfig(..)
  , SolverType(..)
  , parseArgs
  , main
  ) where

import Data.Char (isDigit)
import Data.List (intercalate)
import qualified Data.Map.Strict as Map
import System.Environment (getArgs)
import System.Exit (exitFailure)
import System.IO (hPutStrLn, stderr)

-- | Representation of a single move.
data Move = Move
  { moveDisk :: !Int
  , moveFrom :: !String
  , moveTo :: !String
  }
  deriving (Eq, Ord, Show)

type PegState = Map.Map String [Int]

-- | Render a move in the classic textual form.
renderMove :: Move -> String
renderMove Move {..} =
  "Move disk "
    <> show moveDisk
    <> " from "
    <> moveFrom
    <> " to "
    <> moveTo

-- | Recursive Towers of Hanoi solver producing a lazy move list.
towersOfHanoi :: Int -> String -> String -> String -> [Move]
towersOfHanoi disks source target auxiliary
  | disks <= 0 = []
  | otherwise =
      towersOfHanoi (disks - 1) source auxiliary target
        ++ [Move disks source target]
        ++ towersOfHanoi (disks - 1) auxiliary target source

-- | Iterative solver that emulates recursion via an explicit stack.
towersOfHanoiIterative :: Int -> String -> String -> String -> [Move]
towersOfHanoiIterative disks source target auxiliary = go [(disks, source, target, auxiliary, False)] []
  where
    go [] acc = reverse acc
    go ((n, src, tgt, aux, _stage) : rest) acc | n <= 0 = go rest acc
    go ((n, src, tgt, aux, False) : rest) acc =
      go ((n, src, tgt, aux, True) : (n - 1, src, aux, tgt, False) : rest) acc
    go ((n, src, tgt, aux, True) : rest) acc =
      go ((n - 1, aux, tgt, src, False) : rest) (Move n src tgt : acc)

-- | Produce peg states for visualization, starting with the initial layout.
hanoiStateGenerator :: Int -> String -> String -> String -> [PegState]
hanoiStateGenerator numDisks source target auxiliary =
  let initial =
        Map.fromList
          [ (source, [1 .. numDisks])
          , (auxiliary, [])
          , (target, [])
          ]
   in initial : snd (foldl applyMove (initial, [initial]) moves)
  where
    moves = towersOfHanoi numDisks source target auxiliary

    applyMove (state, acc) Move {..} =
      let disk : rest = Map.findWithDefault [] moveFrom state
          updated =
            Map.insert moveFrom rest $
              Map.insert moveTo (disk : Map.findWithDefault [] moveTo state) state
       in (updated, acc ++ [updated])

-- | Supported solver strategies.
data SolverType = Recursive | Iterative
  deriving (Eq, Show)

-- | Runtime configuration parsed from the CLI.
data CLIConfig = CLIConfig
  { cfgDisks :: !Int
  , cfgSolver :: !SolverType
  , cfgJSON :: !Bool
  , cfgTrace :: !Bool
  , cfgCountOnly :: !Bool
  , cfgMaxSteps :: !(Maybe Int)
  }
  deriving (Eq, Show)

-- | Command-line argument parsing with helpful error reporting.
parseArgs :: [String] -> Either String CLIConfig
parseArgs args = finalize =<< go args defaultState
  where
    defaultState = CLIConfigState Nothing Recursive False False False Nothing

    finalize (CLIConfigState Nothing _ _ _ _ _) = Left "Missing required --disks/-n value"
    finalize (CLIConfigState (Just n) solver json trace countOnly maxSteps)
      | n < 1 = Left "disks must be >= 1"
      | maybe False (<= 0) maxSteps = Left "--max-steps must be positive"
      | otherwise =
          Right
            CLIConfig
              { cfgDisks = n
              , cfgSolver = solver
              , cfgJSON = json
              , cfgTrace = trace
              , cfgCountOnly = countOnly
              , cfgMaxSteps = maxSteps
              }

    go [] st = Right st
    go (flag : rest) st =
      case flag of
        "--help" -> Left usage
        "-h" -> Left usage
        "--disks" -> pickNumber rest $ \n -> go (drop 1 rest) (st {disks = Just n})
        "-n" -> pickNumber rest $ \n -> go (drop 1 rest) (st {disks = Just n})
        "--iterative" -> go rest (st {solver = Iterative})
        "--recursive" -> go rest (st {solver = Recursive})
        "--json" -> go rest (st {json = True})
        "--trace" -> go rest (st {trace = True})
        "--count-only" -> go rest (st {countOnly = True})
        "--max-steps" -> pickNumber rest $ \n -> go (drop 1 rest) (st {maxSteps = Just n})
        _ -> Left $ "Unrecognized argument: " <> flag <> "\n" <> usage

    pickNumber xs k =
      case xs of
        (val : _) | all isDigit val -> k (read val)
        (val : _) -> Left $ "Expected numeric value, got " <> val
        [] -> Left "Expected value after numeric flag"

    usage = unlines
      [ "Usage: ToH [--disks N] [--iterative|--recursive] [--json] [--trace]"
      , "            [--count-only] [--max-steps N]"
      , "  --disks/-n N     Number of disks (required, >= 1)"
      , "  --iterative      Use the stack-based solver"
      , "  --recursive      Use the recursive solver (default)"
      , "  --json           Emit JSON summary instead of plain text"
      , "  --trace          Print peg states after each move"
      , "  --count-only     Only report the minimal move count"
      , "  --max-steps N    Limit the number of moves emitted"
      ]

data CLIConfigState = CLIConfigState
  { disks :: Maybe Int
  , solver :: SolverType
  , json :: Bool
  , trace :: Bool
  , countOnly :: Bool
  , maxSteps :: Maybe Int
  }
  deriving (Eq, Show)

-- | Clamp the emitted move list when --max-steps is provided.
capMoves :: Maybe Int -> [Move] -> [Move]
capMoves Nothing moves = moves
capMoves (Just limit) moves = take limit moves

-- | Render peg states in a compact textual representation.
renderState :: PegState -> String
renderState state =
  let entries =
        [ peg <> ":" <> show (reverse disks)
        | (peg, disks) <- Map.toList state
        ]
   in intercalate " " entries

-- | Manually emit JSON to avoid introducing external dependencies.
renderJSON :: CLIConfig -> Int -> [Move] -> Maybe [PegState] -> String
renderJSON CLIConfig {..} totalMoves moves mStates =
  let moveObjs =
        [ "    {\"disk\":"
            <> show moveDisk
            <> ",\"from\":\""
            <> moveFrom
            <> "\",\"to\":\""
            <> moveTo
            <> "\"}"
        | Move {..} <- moves
        ]
      moveArray =
        case moveObjs of
          [] -> "[]"
          _ -> "[\n" <> intercalate ",\n" moveObjs <> "\n  ]"
      statesArray =
        case mStates of
          Nothing -> "null"
          Just states ->
            case states of
              [] -> "[]"
              _ ->
                "[\n"
                  <> intercalate ",\n"
                    [ "    {"
                        <> intercalate ", "
                          [ "\"" <> peg <> "\": " <> show (reverse stack)
                          | (peg, stack) <- Map.toList st
                          ]
                        <> "}"
                    | st <- states
                    ]
                  <> "\n  ]"
   in "{\n"
        <> "  \"disks\": "
        <> show cfgDisks
        <> ",\n  \"solver\": \""
        <> (case cfgSolver of Recursive -> "recursive"; Iterative -> "iterative")
        <> "\",\n  \"total_moves\": "
        <> show totalMoves
        <> ",\n  \"emitted_moves\": "
        <> show (length moves)
        <> ",\n  \"moves\": "
        <> moveArray
        <> ",\n  \"states\": "
        <> statesArray
        <> "\n}"

-- | Entry point mirroring the Python utility's behaviour.
main :: IO ()
main = do
  args <- getArgs
  case parseArgs args of
    Left msg -> do
      hPutStrLn stderr msg
      exitFailure
    Right cfg@CLIConfig {..} -> do
      let solverFn = case cfgSolver of
            Recursive -> towersOfHanoi
            Iterative -> towersOfHanoiIterative
          allMoves = solverFn cfgDisks "A" "C" "B"
          moves = if cfgCountOnly then [] else capMoves cfgMaxSteps allMoves
          totalMoves = (2 ^ cfgDisks) - 1
          tracedStates =
            if cfgTrace
              then Just $ capStates cfgMaxSteps $ hanoiStateGenerator cfgDisks "A" "C" "B"
              else Nothing
      mapM_ (hPutStrLn stderr . renderState) (maybe [] id tracedStates)
      if cfgJSON
        then putStrLn $ renderJSON cfg totalMoves moves tracedStates
        else if cfgCountOnly
          then putStrLn $ "Total moves: " <> show totalMoves
          else do
            mapM_ putStrLn (zipWith annotate [1 :: Int ..] moves)
            putStrLn $ "\nTotal moves required: " <> show totalMoves
  where
    annotate idx mv = show idx <> ". " <> renderMove mv

    capStates :: Maybe Int -> [PegState] -> [PegState]
    capStates Nothing states = states
    capStates (Just limit) states = take (limit + 1) states
